### PR TITLE
Introduce masking class and incorporate in TokenizerMasking

### LINKED ISSUE #380: Masking class and separation of tokenizer_masking
### Is your feature request related to a problem? Please describe.

Currently masking of data happens in tokenizer_masking in batchify_source and batchify_target, and there is not a class currently for masking, which is needed to implement different masking strategies, vary them through training and so on.

### Describe the solution you'd like

The first solution is a bare bones implementation of this. Create a masker class, which implements a couple of simple masking strategies, and is instantiated in multi_stream_data_sampler before the tokenizer is instantiated. This class should take input data, and return the masked data. This needs to occur for both source and target data.

tokenizer_masking should then be adjusted to do only tokenisation, and no longer do masking at the same time.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### LINKED ISSUE #408: RNGs in tokenizers are not properly initalized
### What happened?

we use time.time() to initialize rngs, which makes runs unreproducible

### What are the steps to reproduce the bug?

Run code

### Version

develop

### Platform (OS and architecture)

Linux

### Relevant log output

```shell

```

### Accompanying data

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by shmh40**
## Description

WIP: create and instantiate a Masker class which implements masking strategies and is called in TokenizerMasking to produce masked source and target tokens. This class should allow the implementation of different masking strategies (e.g. random, per healpix cell, inpainting etc.) and should ensure the masking of source and target are properly aligned. 

Current questions:

1. The Masker returns masked source_data, and the mask perm_sel. This perm_sel is then used in tokenizer_masking, outside of the Masker. Perhaps both the source data and the target data should be produced using this Masker class? We could include, in Masker.mask, if target: use ~perm_sel to produce masked target data.
2. I haven't quite worked out how this should link with different streams, if we want to mask a whole stream and predict from another stream.
3. The second masking strategy, "block" is just a placeholder. Ideally this would be implemented with healpix cells. I need to work out/implement how to use the healpix cells in this Masker class too. Just pass them (produced in TokenizerMasking?) and use them? 

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Resolves #380 
Resolves #408 

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [x] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [x] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [x] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/weathergen/datasets/masking.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        "block", MORE TO BE IMPLEMENTED...).
+        TO BE IMPLEMENTED: masking_combination (str): The strategy for combining masking
+        strategies through training (e.g., "sequential").
+        masking_rate_sampling (bool): Whether to sample the masking rate from a distribution.
+    """
+
+    def __init__(
+        self,
+        masking_rate: float,
+        masking_strategy: str = "random",
```
Comment (tjhunter) [inline src/weathergen/datasets/masking.py]: no default args please

File: src/weathergen/datasets/masking.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        self.masking_strategy = masking_strategy
+        self.masking_combination = masking_combination
+        self.masking_rate_sampling = masking_rate_sampling
+
+    def mask(
+        self,
+        tokenized_data: list[torch.Tensor],
+        rng: np.random.Generator,
+        masking_rate: float = None,
+    ) -> tuple[list[torch.Tensor], list[np.ndarray]]:
```
Comment (tjhunter) [inline src/weathergen/datasets/masking.py]: see the PR I am asking you to review.

File: src/weathergen/datasets/masking.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        split_indices = np.cumsum(token_lens)[:-1]
+        perm_sel = np.split(flat_mask, split_indices)
+
+        # Apply the mask to get the source data (where mask is False)
+        source_data = [data[~p] for data, p in zip(tokenized_data, perm_sel, strict=True)]
+
+        return source_data, perm_sel
+
+
+class Masker_archived:
```
Comment (tjhunter) [inline src/weathergen/datasets/masking.py]: CamelCase

**Review by tjhunter — COMMENTED** [review]
a few style comments

File: src/weathergen/datasets/masking.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    ):
+        self.masking_rate = masking_rate
+        self.masking_strategy = masking_strategy
+        # self.masking_combination = masking_combination
+        self.masking_rate_sampling = masking_rate_sampling
+
+    def mask_source(
+        self,
+        tokenized_data: list[torch.Tensor],
+        rng: np.random.Generator,
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: This should be part of the state, potentially passed in the constructor

File: src/weathergen/datasets/masking.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        masking_rate: float,
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: It's passed in the constructor. Why is it passed here again? any sampling should also happen in this class

File: src/weathergen/datasets/masking.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
+import numpy as np
+import torch
+
+
+class Masker:
+    """Class to generate boolean masks for token sequences and apply them.
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: I think the class can also be used for BERT-type masking + noising.

File: src/weathergen/datasets/masking.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+                                               out and will become the target.
+        """
+        token_lens = [len(t) for t in tokenized_data]
+        num_tokens = sum(token_lens)
+
+        # If there are no tokens, return empty lists.
+        if num_tokens == 0:
+            return tokenized_data, []
+
+        # Determine the masking rate to use for this call
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: Remove, see above

File: src/weathergen/datasets/masking.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        rate = self.masking_rate if masking_rate is None else masking_rate
+
+        if self.masking_rate_sampling:
+            rate = np.clip(
+                np.abs(rng.normal(loc=rate, scale=1.0 / (2.5 * np.pi))),
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: We should parametrize this. But better to do it in a separate PR

File: src/weathergen/datasets/masking.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+            flat_mask = np.zeros(num_tokens, dtype=bool)
+            block_size = int(np.round(rate * num_tokens))
+            if block_size > 0 and num_tokens > 0:
+                start_index = rng.integers(0, max(1, num_tokens - block_size + 1))
+                flat_mask[start_index : start_index + block_size] = True
+        else:
+            assert False, f"Unknown masking strategy: {self.masking_strategy}"
+
+        # Ensure the mask is not degenerate (all True or all False) if there's more than one token.
+        if num_tokens > 1:
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: This should not be necessary. We should see where things break when the corner cases are hit. Maybe this has already been fixed. Otherwise it's also something for a follow-up PR.

File: src/weathergen/datasets/masking.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+            if flat_mask.all():
+                flat_mask[rng.integers(low=0, high=num_tokens)] = False
+            elif not flat_mask.any():
+                flat_mask[rng.integers(low=0, high=num_tokens)] = True
+
+        # Split the flat mask to match the structure of the tokenized data (list of lists)
+        # This will be our perm_sel
+        split_indices = np.cumsum(token_lens)[:-1]
+        perm_sel = np.split(flat_mask, split_indices)
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: Keep perm_sel internal to the class

File: src/weathergen/datasets/masking.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+                processed_target_tokens.append(torch.cat(selected_tensors))
+            else:
+                processed_target_tokens.append(
+                    torch.empty(0, feature_dim, dtype=coords.dtype, device=coords.device)
+                )
+
+        return processed_target_tokens
+
+
+class MaskerArchived:
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: What is this class about?

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 12]
```diff
@@ -336,6 +339,8 @@ def __iter__(self):
 
                             (ss_cells, ss_lens, ss_centroids) = self.tokenizer.batchify_source(
                                 stream_info,
+                                # NOTE: two unused arguments in TokenizerMasking,
+                                # still used in TokenizerForecast?
                                 self.masking_rate,
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: No, they should be removed

File: src/weathergen/datasets/tokenizer_utils.py [hunk 13]
```diff
@@ -178,6 +178,7 @@ def tokenize_window_space(
     if len(source) < 2:
         return
 
+    # idx_ord_lens is length...
```
Comment (clessig) [inline src/weathergen/datasets/tokenizer_utils.py]: # idx_ord_lens is length is number of tokens per healpix cell

File: src/weathergen/datasets/tokenizer_masking.py [hunk 14]
```diff
@@ -289,16 +286,40 @@ def id(arg):
...
-        # TODO: implement sampling rate target
-        target_tokens = [
-            (
-                torch.cat([c if p else torch.tensor([]) for c, p in zip(cc, pp, strict=True)])
-                if len(cc) > 0
-                else torch.tensor(cc)
-            )
-            for cc, pp in zip(target_tokens_cells, self.perm_sel, strict=True)
-        ]
+        ######################
```
Comment (clessig) [inline src/weathergen/datasets/tokenizer_masking.py]: Remove before we merge

File: src/weathergen/datasets/tokenizer_masking.py [hunk 15]
```diff
@@ -280,7 +278,6 @@ def id(arg):
         )
 
         # tokenize
-        # TODO: properly set stream_id; don't forget to normalize
```
Comment (clessig) [inline src/weathergen/datasets/tokenizer_masking.py:283]: This still needs to be done

File: src/weathergen/datasets/tokenizer_masking.py [hunk 16]
```diff
@@ -177,46 +171,49 @@ def batchify_source(
...
+        # NOTE: Is it already handled in the masker?
 
-        # ensure that masking is not degenerate i.e. it is not all true or false
-        if not mask.any():
-            mask[self.rng.integers(low=0, high=len(mask))] = True
-        if mask.all():
-            mask[self.rng.integers(low=0, high=len(mask))] = False
+        # if masking rate is 1.0, all tokens are masked, so the source is empty
+        # but we must compute perm_sel for the target function
+        if masking_rate == 1.0:
```
Comment (clessig) [inline src/weathergen/datasets/tokenizer_masking.py]: This case should be handled in the Masker and not here

File: src/weathergen/datasets/masking.py (hunk 4 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py]: Agreed thank you - done with new commit

File: src/weathergen/datasets/masking.py (hunk 5 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py]: Agreed

File: src/weathergen/datasets/masking.py (hunk 6 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py]: Updated the docstring, remove boolean

File: src/weathergen/datasets/masking.py (hunk 7 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py]: Done

File: src/weathergen/datasets/masking.py (hunk 10 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py]: Sorry, done

File: src/weathergen/datasets/masking.py (hunk 11 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py]: Older modified version of the class for my reference, now removed

File: src/weathergen/datasets/masking.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    def __init__(
+        self,
+        masking_rate: float,
+        masking_strategy: str,
+        # masking_combination: str,
+        masking_rate_sampling: bool,
+    ):
+        self.masking_rate = masking_rate
+        self.masking_strategy = masking_strategy
+        # self.masking_combination = masking_combination
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: Can we remove this line before we merge.

File: src/weathergen/datasets/masking.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+        self.masking_rate_sampling = masking_rate_sampling
+        
+        # Initialize the random number generator.
+        worker_info = torch.utils.data.get_worker_info()
+        div_factor = (worker_info.id + 1) if worker_info is not None else 1        
+        self.rng = np.random.default_rng(int(time.time() / div_factor))
```
Comment (clessig) [inline src/weathergen/datasets/masking.py:36]: The rng seed should be passed from the cf.seed, to ensure we can reproduce maskings if we want, while ensuring it is different for the different parallel workers. 

I would suggest to keep it as is to not overload the PR but open once merged, open an issue to address the problem.

File: src/weathergen/datasets/masking.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+        # If masking_rate_sampling is enabled, sample the rate from a normal distribution.
+        if self.masking_rate_sampling:
+            rate = np.clip(
+                np.abs(self.rng.normal(loc=rate, scale=1.0 / (2.5 * np.pi))),
+                0.0,
+                1.0,
+            )
+
+        # Handle the special case where all tokens are masked
+        # NOTE: not going to handle different streams correctly.
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: I don't fully understand the comment. Can you please elaborate.

File: src/weathergen/datasets/masking.py [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+            source_data = [data[~p] for data, p in zip(tokenized_data, self.perm_sel, strict=True)]
+            return source_data
+
+        # Implementation of different masking strategies.
+        # Generate a flat boolean mask based on the strategy.
+
+        if self.masking_strategy == "random":
+            flat_mask = self.rng.uniform(0, 1, num_tokens) < rate
+
+        elif self.masking_strategy == "block":
```
Comment (clessig) [inline src/weathergen/datasets/masking.py:95]: Have you visualized this and checked correctness?

File: src/weathergen/datasets/masking.py [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+            coords (torch.Tensor): Coordinates tensor, used to determine feature dimension.
+            geoinfos (torch.Tensor): Geoinfos tensor, used to determine feature dimension.
+            source (torch.Tensor): Source tensor, used to determine feature dimension.
+
+        Returns:
+            list[torch.Tensor]: The target data with masked tokens, one tensor per cell.
+        """
+
+        # check that self.perm_sel is set with an assert statement
+        assert hasattr(self, 'perm_sel'), "Masker.perm_sel must be set (in mask_source) before calling mask_target."
```
Comment (clessig) [inline src/weathergen/datasets/masking.py]: You set it in the constructor (to None) so the hasattr will alway be true. You need to test for not-None

File: src/weathergen/datasets/masking.py (hunk 20 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py:95]: Yes, visualised and looks as expected. Just a placeholder for now. Healpix in next PR.

File: src/weathergen/datasets/masking.py (hunk 19 above)
Comment (shmh40) [reply src/weathergen/datasets/masking.py]: Removed, my misunderstanding

**Comment by shmh40** [comment]
Note, this PR does introduce a new config parameter, and hence other developers will be notified.

# include a masking strategy here, currently only supporting "random" and "block"
masking_strategy: "random"
